keyboard_handler/include/keyboard_handler/keyboard_handler_base.hpp

```diff
@@ -0,0 +1,300 @@
+  static constexpr callback_handle_t invalid_handle = 0;
+
+  /// \brief Adding callable object as a handler for specified key press combination.
+  /// \param callback Callable which will be called when key_code will be recognized.
+  /// \param key_code Value from enum which corresponds to some predefined key press combination.
+  /// \return Return Newly created callback handle if callback was successfully added to the
+  /// keyboard handler, returns invalid_handle if callback is nullptr or keyboard handler wasn't
+  /// successfully initialized.
+  KEYBOARD_HANDLER_PUBLIC
+  callback_handle_t add_key_press_callback(
+    const callback_t & callback,
+    KeyboardHandlerBase::KeyCode key_code);
+
+  /// \brief Delete callback from keyboard handler callback's list
+  /// \param handle Callback's handle returned from #add_key_press_callback
+  KEYBOARD_HANDLER_PUBLIC
+  void delete_key_press_callback(const callback_handle_t & handle) noexcept;
+
+protected:
+  struct callback_data
+  {
+    callback_handle_t handle;
+    callback_t callback;
+  };
+  bool is_init_succeed_ = false;
+  std::mutex callbacks_mutex_;
+  std::unordered_multimap<KeyCode, callback_data> callbacks_;
+
+private:
+  callback_handle_t get_new_handle();
+  callback_handle_t last_handle_ = 0;
+};
+
+enum class KeyboardHandlerBase::KeyCode: uint32_t
+{
+  UNKNOWN = 0,
+  NUMBER_1,
+  NUMBER_2,
+  NUMBER_3,
+  NUMBER_4,
+  NUMBER_5,
+  NUMBER_6,
+  NUMBER_7,
+  NUMBER_8,
+  NUMBER_9,
+  NUMBER_0,
+  MINUS,
+  EQUAL_SIGN,
+  A,
+  B,
+  C,
+  D,
+  E,
+  F,
+  G,
+  H,
+  I,
+  J,
+  K,
+  L,
+  M,
+  N,
+  O,
+  P,
+  Q,
+  R,
+  S,
+  T,
+  U,
+  V,
+  W,
+  X,
+  Y,
+  Z,
+  CAPITAL_A,
+  CAPITAL_B,
+  CAPITAL_C,
+  CAPITAL_D,
+  CAPITAL_E,
+  CAPITAL_F,
+  CAPITAL_G,
+  CAPITAL_H,
+  CAPITAL_I,
+  CAPITAL_J,
+  CAPITAL_K,
+  CAPITAL_L,
+  CAPITAL_M,
+  CAPITAL_N,
+  CAPITAL_O,
+  CAPITAL_P,
+  CAPITAL_Q,
+  CAPITAL_R,
+  CAPITAL_S,
+  CAPITAL_T,
+  CAPITAL_U,
+  CAPITAL_V,
+  CAPITAL_W,
+  CAPITAL_X,
+  CAPITAL_Y,
+  CAPITAL_Z,
+  CURSOR_UP,
+  CURSOR_DOWN,
+  CURSOR_LEFT,
+  CURSOR_RIGHT,
+  ESCAPE,
+  SPACE,
+  ENTER,
+  BACK_SPACE,
+  DELETE,
+  END,
+  PG_DOWN,
+  PG_UP,
+  HOME,
+  INSERT,
+  F1,
+  F2,
+  F3,
+  F4,
+  F5,
+  F6,
+  F7,
+  F8,
+  F9,
+  F10,
+  F11,
+  F12,
+  SHIFT_F1,
+  SHIFT_F2,
+  SHIFT_F3,
+  SHIFT_F4,
+  SHIFT_F5,
+  SHIFT_F6,
+  SHIFT_F7,
+  SHIFT_F8,
+  SHIFT_F9,
+  SHIFT_F10,
+  SHIFT_F11,
+  SHIFT_F12,
+  END_OF_KEY_CODE_ENUM
+};
+
+/// \brief Prefix increment operator for KeyCode enum values
+KEYBOARD_HANDLER_PUBLIC
+KeyboardHandlerBase::KeyCode & operator++(KeyboardHandlerBase::KeyCode & key_code);
+
+/// \brief Data type for mapping KeyCode enum value to it's string representation.
+struct KeyCodeToStrMap
+{
+  KeyboardHandlerBase::KeyCode inner_code;
+  const char * str;
+};
+
+/// \brief Lookup table for mapping KeyCode enum value to it's string representation.
+static const KeyCodeToStrMap ENUM_KEY_TO_STR_MAP[] {
```

Comment:
Maybe we would be better off with something less manual, like https://github.com/aantron/better-enums/blob/master/enum.h - this is going to be hard to maintain.

Not blocking for first pass of this.

Reply:
@emersonknapp I am totally fine to move it somewhere in `github.com/ros-tooling/keyboard_handler`. I think it would be a more proper place for it.

As regards to the usage better enums. I am aware about it. Although it has some [limitations](https://github.com/aantron/better-enums#limitations). \
For instance maximum number of elements couldn't be more than **127** when compile with MS Visual C++ and we certainly will need more. 

Though meanwhile, I don't think that it would be difficult to support current implementation. I have added unit test to check if any enum values doesn't have corresponding string map values. \
If someone will add new enum value for key code and will forgot to add string representation in map, those unit test will fail and print out which key code doesn't have string value in map.

Reply:
That's fine for now. I'm more interested in getting the minimal amount of functionality so support what rosbag2 needs, which is simple un-modified keypresses.